server/Arcadia.Assistant.ExternalStorages.SharepointOnline/SharepointAuthTokenService.cs

```diff
@@ -0,0 +1,208 @@
+            var hostName = new Uri(sharepointUrl).Authority;
+
+            var oauthClientId = this.GetFormattedPrincipal(this.configuration.ClientId, null, realm);
+            var oauthResource = this.GetFormattedPrincipal(SharepointPrincipal, hostName, realm);
+
+            var accessToken = await this.IssueToken(
+                realm,
+                oauthClientId,
+                this.configuration.ClientSecret,
+                oauthResource,
+                cancellationToken);
+            return accessToken;
+        }
+
+        private async Task<string> GetRealm(string sharepointUrl, CancellationToken cancellationToken)
+        {
+            using (var httpClient = this.httpClientFactory.CreateClient())
+            {
+                var request = new HttpRequestMessage(HttpMethod.Get, sharepointUrl + "/_vti_bin/client.svc");
+                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", string.Empty);
+
+                var response = await httpClient.SendAsync(request, cancellationToken);
+
+                if (response.StatusCode != HttpStatusCode.Unauthorized || response.Headers.WwwAuthenticate == null)
+                {
+                    throw new HttpRequestException("Not supported response to a realm request");
+                }
+
+                var wwwAuthenticateHeaderValue = response.Headers.WwwAuthenticate.ToString();
+
+                var realm = this.GetRealmFromHeader(wwwAuthenticateHeaderValue);
+
+                if (realm == null)
+                {
+                    throw new HttpRequestException("Not supported response to a realm request");
+                }
+
+                return realm;
+            }
+        }
+
+        private string GetRealmFromHeader(string headerValue)
+        {
+            const string bearer = "Bearer realm=\"";
+            const int realmLength = 36;
+
+            var bearerIndex = headerValue.IndexOf(bearer, StringComparison.Ordinal);
+
+            if (bearerIndex < 0)
+            {
+                return null;
+            }
+
+            var realmStartIndex = bearerIndex + bearer.Length;
+
+            if (headerValue.Length < realmStartIndex + realmLength)
+            {
+                return null;
+            }
+
+            var realm = headerValue.Substring(realmStartIndex, realmLength);
+
+            return !Guid.TryParse(realm, out _) ? null : realm;
+        }
+
+        private async Task<string> IssueToken(
+            string realm,
+            string tokenClientId,
+            string tokenClientSecret,
+            string tokenResource,
+            CancellationToken cancellationToken)
+        {
+            using (var httpClient = this.httpClientFactory.CreateClient())
+            {
+                var tokenUrl = await this.GetOAuth2Url(realm, cancellationToken);
+
+                var requestData =
+                    "grant_type=" + HttpUtility.UrlEncode(OAuth2GrantType) +
```

Comment:
uri builder?

Reply:
No, because this is not URI, this is request body.